### PR TITLE
L2 Normalization and Triplet loss functions

### PR DESCRIPTION
**Pull request by amitibo**
This PR adds two functions:
1) l2_normalization: this function normalizes a 1D vector by its l2 norm.
2) triplet: this function implements a triplet loss as described in 'Learning Fine-grained Image Similarity with Deep Ranking`

### CODE REVIEW

**Comment by unnonouno** [comment]
Is it possible to split two functions with two PRs?

File: chainer/functions/loss/triplet.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+from chainer.utils import type_check
+
+
+class Triplet(function.Function):
+
+    """Triplet loss function."""
+
+    def __init__(self, margin):
+        if margin <= 0:
+            raise ValueError("margin should be positive value.")
```
Comment (unnonouno) [inline chainer/functions/loss/triplet.py]: Use single quotes.

File: chainer/functions/loss/triplet.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            in_types[0].shape[0] > 0
+        )
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+
+        a, p, n = inputs  # anchor, positive, negative
+        N = a.shape[0]
+
+        dist = xp.sum((a-p)**2 - (a-n)**2, axis=1) + self.margin
```
Comment (unnonouno) [inline chainer/functions/loss/triplet.py]: Please insert spaces at the side of operators.

File: chainer/functions/loss/triplet.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        mask = xp.array(tmp > 0, dtype=numpy.float32)
+
+        gx0 = (2 * gy[0] * (n - p) * mask / N).astype(numpy.float32)
+        gx1 = (2 * gy[0] * (p - a) * mask / N).astype(numpy.float32)
+        gx2 = (2 * gy[0] * (a - n) * mask / N).astype(numpy.float32)
+
+        return gx0, gx1, gx2
+
+
+def triplet(a, p, n, margin=0.2):
```
Comment (unnonouno) [inline chainer/functions/loss/triplet.py]: Use `anchor`, `positive`, and `negative` for names of variables for readability.

File: chainer/functions/normalization/l2_normalization.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        x_type, = in_types
+
+        type_check.expect(
+            x_type.dtype == numpy.float32,
+            x_type.ndim == 2,
+        )
+
+    def forward_cpu(self, inputs):
+        x = array.as_mat(inputs[0])
+        norm = numpy.sqrt((x * x).sum(axis=1)) + self.eps
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Can you use `numpy.linalg.norm`?

File: chainer/functions/normalization/l2_normalization.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+        x = inputs[0]
+        gy = gy[0]
+
+        xp = cuda.get_array_module(x)
+
+        if xp is numpy:
+            norm = numpy.sqrt((x * x).sum(axis=1)) + self.eps
+            norm = norm[:, numpy.newaxis]
+
+            gx = gy*norm - (x*gy).sum(axis=1)[:, numpy.newaxis]*x/norm
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Insert spaces.

File: chainer/functions/normalization/l2_normalization.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                x.shape
+            )
+            x_gy = cuda.cupy.broadcast_to(
+                (x*gy).sum(axis=1, keepdims=True),
+                x.shape
+            )
+            gx = cuda.elementwise(
+                'T gy, T x, T x_gy, T norm',
+                'T gx',
+                'gx = (gy*norm - x_gy*x/norm)/(norm*norm)',
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Insert spaces.

File: chainer/functions/__init__.py [hunk 7]
```diff
@@ -161,6 +163,8 @@
 BinaryHierarchicalSoftmax = hierarchical_softmax.BinaryHierarchicalSoftmax
 Contrastive = contrastive.Contrastive
 contrastive = contrastive.contrastive
+Triplet = triplet.Triplet
+triplet = triplet.triplet
```
Comment (unnonouno) [inline chainer/functions/__init__.py]: Please sort alphabetical order. I mean, please move these two lines after `softmax_cross_entropy`.

File: chainer/functions/__init__.py [hunk 8]
```diff
@@ -183,6 +187,8 @@
 batch_inv = inv.batch_inv
 BatchL2NormSquared = batch_l2_norm_squared.BatchL2NormSquared
 batch_l2_norm_squared = batch_l2_norm_squared.batch_l2_norm_squared
+L2Normalization = l2_normalization.L2Normalization
```
Comment (unnonouno) [inline chainer/functions/__init__.py]: Move them after `batch_normalization`.

File: tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+from chainer.utils import type_check
+
+
+@testing.parameterize(
+    {'shape': (4, 15)},
+    {'shape': (4, 15)},
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py]: It looks the same condition?

File: tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    def check_forward(self, x_data):
+        x = chainer.Variable(x_data)
+
+        y = functions.l2_normalization(x)
+        self.assertEqual(y.data.dtype, numpy.float32)
+        y_data = cuda.to_cpu(y.data)
+
+        y_expect = numpy.empty_like(self.x)
+        for n in six.moves.range(len(self.x)):
+            y_expect[n] = self.x[n]/numpy.linalg.norm(self.x[n])
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py]: Insert spaces at the side of `/`.

File: chainer/functions/loss/triplet.py [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+            in_types[0].shape == in_types[1].shape,
+            in_types[0].shape == in_types[2].shape,
+            in_types[0].shape[0] > 0
+        )
+
+    def forward(self, inputs):
+        xp = cuda.get_array_module(*inputs)
+
+        anchor, positive, negative = inputs
+        N = anchor.shape[0]
```
Comment (unnonouno) [inline chainer/functions/loss/triplet.py]: Use `N = len(anchor)`

File: chainer/functions/loss/triplet.py [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+        self.dist_hinge = xp.maximum(dist, 0)
+        loss = xp.sum(self.dist_hinge) / N
+
+        return xp.array(loss, dtype=numpy.float32),
+
+    def backward(self, inputs, gy):
+        xp = cuda.get_array_module(*inputs)
+
+        anchor, positive, negative = inputs
+        N = anchor.shape[0]
```
Comment (unnonouno) [inline chainer/functions/loss/triplet.py]: `N = len(anchor)`

File: chainer/functions/normalization/l2_normalization.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+        x = array.as_mat(inputs[0])
+        norm = numpy.linalg.norm(x, axis=1) + self.eps
+        return x / norm[:, numpy.newaxis],
+
+    def forward_gpu(self, inputs):
+        x = array.as_mat(inputs[0])
+        l2norm_kernel = cuda.cupy.ReductionKernel(
+            'T x', 'T y', 'x * x', 'a + b', 'y = sqrt(a)', '0', 'l2norm'
+        )
+        norm = cuda.cupy.broadcast_to(
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Do you need to add `eps` here?

File: chainer/functions/normalization/l2_normalization.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+            norm = numpy.sqrt((x * x).sum(axis=1)) + self.eps
+            norm = norm[:, numpy.newaxis]
+
+            gx = gy * norm - (x * gy).sum(axis=1)[:, numpy.newaxis] * x / norm
+            gx = gx / norm**2
+        else:
+            l2norm_kernel = cuda.cupy.ReductionKernel(
+                'T x', 'T y', 'x * x', 'a + b', 'y = sqrt(a)', '0', 'l2norm'
+            )
+            norm = cuda.cupy.broadcast_to(
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Is `eps` required?

File: chainer/functions/normalization/l2_normalization.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+                l2norm_kernel(x, axis=1).reshape(-1, 1),
+                x.shape
+            )
+            x_gy = cuda.cupy.broadcast_to(
+                (x*gy).sum(axis=1, keepdims=True),
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Insert spaces at the side of `*`.

File: chainer/functions/normalization/l2_normalization.py [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    """L2 norm (a.k.a. Euclidean norm) squared.
+
+    This function implements L2 normalization on a 1D vector. No reduction
+    along batch axis is done.  Let :math:`x` be an input vector of dimension
+    :math:`(N,D)`, where :math:`N` and :math:`K` denote mini-batchsize and the
+    dimension of the input variable. Then, this function computes an output
+    vector :math:`y` by following formula:
+
+    .. math::
+       y_i = {x_i \\over \\| x_i \\|_2}
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: Can you write effect of `eps`?

File: tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+from chainer.utils import type_check
+
+
+@testing.parameterize(
+    {'shape': (4, 15)},
+    {'shape': (4, 15)},
+)
+class TestL2Normalization(unittest.TestCase):
+
+    def setUp(self):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py]: Please add a test case when `x` is very small. I mean, write a test for `eps`.

File: chainer/functions/normalization/l2_normalization.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+            gx = cuda.elementwise(
+                'T gy, T x, T x_gy, T norm',
+                'T gx',
+                'gx = (gy * norm - x_gy * x / norm) / (norm * norm)',
+                'l2_bwd')(gy, x, x_gy, norm)
+
+        return gx,
+
+
+def l2_normalization(x, eps=1e-5):
```
Comment (unnonouno) [inline chainer/functions/normalization/l2_normalization.py]: I think `normalize` is better because scikit-learn has `normalize` function. We want to use the same name as the existing libraries such as numpy.
http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html

And use `NormalizeL2` as its class name.

**Comment by amitibo** [comment]
@unnonouno Sorry for the newbie question: I seemed to have mixed several branches into my fork ``master`` branch. I am wondering what is the preferred forking/pull request workflow: Having ``master`` branch a mirror of the main repository ``master`` branch, developing on new branches (e.g. ``l2_normalize`` and ``triplet``) and opening pull requests from these branches? If so, should I delete my fork and start anew, including opening a new pull request?

A second question. I am not sure what is the correct way to implement the ``esp`` functionality in the gpu version of l2_normalization. Is it another input to the cuda kernel, and can a scalar input be used along side a vector input?

Thanks.

**Comment by unnonouno** [comment]
Yes, I recommend you to make a new branch for each feature and each PR. We can merge PRs one by one.
BTW, this PR is OK. I'll merge it when both fixes become acceptable.

And, you can send a scalar value to CUDA kernel as its input. See this implementation https://github.com/pfnet/chainer/blob/master/chainer/functions/loss/negative_sampling.py#L78

**Comment by unnonouno** [comment]
I'll close this PR because both #1330 and #1331 are merged. Thank you very much!
